cmd/update-main.go

```diff
@@ -31,204 +33,254 @@ import (
-const (
-	mcUpdateStableURL       = "https://dl.minio.io/client/mc/release"
-	mcUpdateExperimentalURL = "https://dl.minio.io/client/mc/experimental"
-)
-
 // updateMessage container to hold update messages.
 type updateMessage struct {
-	Status   string `json:"status"`
-	Update   bool   `json:"update"`
-	Download string `json:"downloadURL"`
-	Version  string `json:"version"`
+	Status    string `json:"status"`
+	Update    bool   `json:"update"`
+	Download  string `json:"downloadURL,omitempty"`
+	Version   string `json:"version"`
+	olderThan time.Duration
 }
 
 // String colorized update message.
 func (u updateMessage) String() string {
-	if !u.Update {
+	if u.olderThan == time.Duration(0) {
 		return console.Colorize("Update", "You are already running the most recent version of ‘mc’.")
 	}
-	var msg string
-	if runtime.GOOS == "windows" {
-		msg = "Download " + u.Download
-	} else {
-		msg = "Download " + u.Download
-	}
-	msg, err := colorizeUpdateMessage(msg)
-	fatalIf(err.Trace(msg), "Unable to colorize experimental update notification string ‘"+msg+"’.")
-	return msg
+	return colorizeUpdateMessage(u.Download, u.olderThan)
 }
 
 // JSON jsonified update message.
 func (u updateMessage) JSON() string {
 	u.Status = "success"
+	u.Update = u.olderThan != time.Duration(0)
 	updateMessageJSONBytes, e := json.Marshal(u)
 	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
 
 	return string(updateMessageJSONBytes)
 }
 
-func parseReleaseData(data string) (time.Time, *probe.Error) {
-	releaseStr := strings.Fields(data)
-	if len(releaseStr) < 2 {
-		return time.Time{}, probe.NewError(errors.New("Update data malformed"))
+const releaseTagTimeLayout = "2006-01-02T15-04-05Z"
+const mcReleaseURL = "https://dl.minio.io/client/mc/release/" + runtime.GOOS + "-" + runtime.GOARCH + "/"
+
+func getCurrentReleaseTime(mcVersion, mcBinaryPath string) (releaseTime time.Time, err *probe.Error) {
+	var e error
+	if releaseTime, e = time.Parse(time.RFC3339, mcVersion); e == nil {
+		return releaseTime, probe.NewError(e)
 	}
-	releaseDate := releaseStr[1]
-	releaseDateSplits := strings.SplitN(releaseDate, ".", 3)
-	if len(releaseDateSplits) < 3 {
-		return time.Time{}, probe.NewError(errors.New("Update data malformed"))
+
+	// Looks like version is mc non-standard, we use mc binary's ModTime as release time.
+	fi, e := os.Stat(mcBinaryPath)
+	if e != nil {
+		e = fmt.Errorf("Unable to get ModTime of %s. %s", mcBinaryPath, e)
+	} else {
+		releaseTime = fi.ModTime().UTC()
 	}
-	if releaseDateSplits[0] != "mc" {
-		return time.Time{}, probe.NewError(errors.New("Update data malformed, missing mc tag"))
+
+	return releaseTime, probe.NewError(e)
+}
+
+// GetCurrentReleaseTime - returns this process's release time.  If it is official minio version,
+// parsed version is returned else minio binary's mod time is returned.
+func GetCurrentReleaseTime() (releaseTime time.Time, err *probe.Error) {
+	return getCurrentReleaseTime(Version, os.Args[0])
```

Comment:
os.Args[0] is sometimes 'mc' if the binary is in one of PATH directories, os.Stat() would fail because we are not providing it an absolute path. With go 1.8, there is os.Executable() which helps.